gprofiler/profilers/java.py

```diff
@@ -261,36 +261,22 @@ def _is_musl(self) -> bool:
 
     def _copy_libap(self) -> None:
         # copy *is* racy with respect to other processes running in the same namespace, because they all use
-        # the same directory for libasyncProfiler.so, as we don't want to create too many copies of it that
-        # will waste disk space.
-        # my attempt here is to produce a race-free way of ensuring libasyncProfiler.so was fully copied
-        # to a *single* path, per namespace.
-        # newer kernels (>3.15 for ext4) have the renameat2 syscall, with RENAME_NOREPLACE, which lets you
-        # atomically move a file without replacing if the target already exists.
-        # this function operates similarly on directories. if you rename(dir_a, dir_b) then dir_b is replaced
-        # with dir_a iff it's empty. this gives us the same semantics.
-        # so, we create a temporary directory on the same filesystem (so move is atomic) in a race-free way
-        # by including the PID in its name; then we move it.
-        # TODO: if we ever move away from the multithreaded model, we can get rid of this complexity
-        # by ensuring a known order of execution.
-        ap_dir_host_tmp = f"{self._ap_dir_host}.{self.process.pid}"
-        os.makedirs(ap_dir_host_tmp)
-        libap_tmp = os.path.join(ap_dir_host_tmp, "libasyncProfiler.so")
-        shutil.copy(
-            resource_path(os.path.join("java", "musl" if self._is_musl() else "glibc", "libasyncProfiler.so")),
-            libap_tmp,
-        )
-        os.chmod(libap_tmp, 0o755)  # make it accessible for all; needed with PyInstaller, which extracts files as 0700
-        try:
-            os.rename(ap_dir_host_tmp, self._ap_dir_host)
-        except OSError as e:
-            if e.errno == errno.ENOTEMPTY:
-                # remove our copy
-                shutil.rmtree(ap_dir_host_tmp)
-                # it should have been created by somene else.
-                assert os.path.exists(self._libap_path_host)
-            else:
-                raise
+        # the same directory for libasyncProfiler.so.
+        # therefore, we need to synchronize copies from different threads that profile different processes.
+        if os.path.exists(self._libap_path_host):
+            # all good
+            return
+
+        with libap_copy_lock:
+            if not os.path.exists(self._libap_path_host):
+                # atomically copy it
+                libap_resource = resource_path(
+                    os.path.join("java", "musl" if self._is_musl() else "glibc", "libasyncProfiler.so")
+                )
+                os.chmod(
+                    libap_resource, 0o755
+                )  # make it accessible for all; needed with PyInstaller, which extracts files as 0700
+                atomic_copy(libap_resource, self._libap_path_host)
```

Comment:
Why do we need an atomic copy? Why not just a regular copy? The lock protects it

Reply:
I recall I had a concrete reason when I wrote it, Too bad I didn't document. Anyway, 2 new reasons from the top of my head - 
1. As a best practice when moving around executables. Linux behaves differently when you `open(.. ,"w")` a file and `os.rename` a file. The former edits the existing inode, the latter creates a new one (and existing references to the old inode are kept).
2. Besides the lock, we do protect this with `os.path.exists` so there's no chance we're `open`ing a file already loaded...

Anyway, this is still safer, when moving exes I strive to always use copy-rename

Reply:
I much prefer the simpler lock and bool (instead of relying on FS exists):
```python
with lock:
    if not has_copied_libap:
        shutil.copy(src, dst)
        has_copied_libap = True
```

Less code, less stuff to wonder about when reading, but OK